src/pudl/metadata/fields.py

```diff
@@ -3680,7 +3685,418 @@
             }
         },
     },
-    "eia": {"fuel_units": {"constraints": {"enum": sorted(FUEL_UNITS_EIA.keys())}}},
+    "eia": {
```

Comment:
screed alert, sorry:

While our current system is "put all fields in one big bucket," I think we should move towards splitting our fields up into meaningful categories.

The benefit of putting everything in one general bucket is that any resource can use these fields without having to redefine them.

The downside is that it makes it harder to add new fields and asset schemas down the line:
- when adding a new field with the same name as an existing one, it's hard to know where the existing one came from & what contexts its used in, so it's hard to make an informed decision about what to do with the new field
- if you are thinking about what to name your column, and want to see what other related assets have named a similar column, you are stuck wading through *all* columns ever named

It's sort of like shoving all of your functions into one module.

We attempted to mitigate these downsides with our group/resource override system, which is what I'm using here, but that system is pretty janky...
* `Resource.dict_from_resource_descriptor` does a bunch of logic to figure out which override to use, but...
*`Field.from_id` doesn't understand overrides directly
*`get_pudl_dtypes` applies group overrides, but not resource overrides

My dream is to evolve our current system into one that follows industry practices such as

1. Bundling the field definitions *with* the schema definitions, like most other schema systems like Frictionless, Pandera, SQLAlchemy, Pydantic etc.
2. Allowing for re-use of fields by direct, explicit reference - also like those above systems.

And I think that shoving these fields into an `"eia"` group is how we stop digging this hole quite so quickly while we wait for time to make the system better. I had initially hoped to shove them into "eia923_cooling" and "eia860_cooling" groups, but our system doesn't support "multiple namespaces" like that...

Reply:
@zaneselvans @cmgosnell : Ella & I talked a bit more about this and have a couple bullet points:

* having One Big Honkin namespace does provide some structural defense against bad naming
* but, if we want to have field defs colocated with schema definitions at some point in the future (which makes defining schemas + resources much easier), we will have to abandon One Big Dictionary at some point

We're curious to see if there's other context about why we have the fields set up the way they are, and what you think about putting these fields in the general dictionary or keeping them separate.

Reply:
I think the main motivation that we had when we set up the One Giant Namespace was avoiding redefining the same columns in many different places, and needing to keep those definitions synchronized, which of course never happens reliably -- this was a problem we had with the system prior to creating the current metadata setup.

I could imagine the opposite convention being fine, where by default each resource has to define its own fields, but it has the option of explicitly grabbing a definition from somewhere else to reuse it if it means the same thing / represents the same information, to avoid duplication and ensure that all the field definitions stay consistent. I do worry that we might not notice when we are giving different fields the same names even when they don't really have the same meaning. But also there are overly generic field names in the current system that should really be differentiated somehow.

I'm apprehensive about starting to do things differently but only with new fields, since it feels like added complexity / internal inconsistency, where we're doing the same thing in multiple different ways simultaneously. But I see that One Giant Namespace doesn't seem like it can scale gracefully to thousands of named columns.

Reply:
> I think the main motivation that we had when we set up the One Giant Namespace was avoiding redefining the same columns in many different places

That was my understanding also, and I think the opposite convention will let us do this while also giving us the benefits of explicit references, namespacing, etc.

> I do worry that we might not notice when we are giving different fields the same names even when they don't really have the same meaning.

Yeah, Ella and I talked a bit about this issue - if a column name is the same across multiple tables, users might expect it to mean the same thing, even if they don't. Sort of like when you shadow a variable. We could run a check on pre-commit that checks for "shadowed" field names. In our schema definitions we could allow a "shadow_ok" flag that says, effectively, "Yeah, I know this shares the same name as something else and I don't care. This is both (a) different from the other thing and (b) not different enough that a different name makes sense"

> I'm apprehensive about starting to do things differently but only with new fields, since it feels like added complexity / internal inconsistency, where we're doing the same thing in multiple different ways simultaneously. But I see that One Giant Namespace doesn't seem like it can scale gracefully to thousands of named columns.

I think "migrating to a new system" here has to be done in one fell swoop, but "preparing to migrate to such a new system" doesn't. That being said, I think we can write a few scripts, re-do a medium of the resource + package logic, and get converted over to the new system without *too* much hassle... I've been meaning to write a little project pitch for that for a while.

Reply:
I'm worried about a lag between "preparing for the transition" and "having funding and time to do the transition" that looks like a mixed system for years, where we're using the group name spaces to mean two slightly different things and our docs don't clearly reflect what our practices are. Without the validation tests that make sure we don't have unexpected "shadowed" variables, I'm also hesitant to start doing this - plus, this PR is already so big that I don't want to start adding more metadata work in here, this feels like scope creep. So I'd lean towards preferring to stick with our current system for now, and moving these back to One Giant Namespace.

Reply:
Yeah, I think that lag is certainly annoying/frustrating. I think that it's easier to combine namespaces than to split them, because the information of "where this field *would* belong" gets lost in the combination, but I'm OK with pushing things into One Giant Namespace (OGN).

Partially because so many things are already in the OGN that whatever we do to deal with it will have to figure out some way to split the namespace up semi-automatically anyway - so what's a few more fields to migrate in that world?